<div style="border:solid Chocolate 2px; padding: 40px">

<b>Привет!👋</b>

Меня зовут Огурлиев Анзор, я буду ревьюером твоего проекта. Сразу хочу предложить тебе общаться на "ты", если тебе это неудобно, скажи мне об этом. Нам предстоит пройти с тобой через тернистый и от того интересный путь анализа данных! Можешь не переживать, я всегда помогу тебе, если у тебя возникнут вопросы)

По ходу твоей работы я буду оставлять свои комментарии, по выполненным тобой шагам. Попрошу тебя не удалять их, так мне будет проще отслеживать твой прогресс в случае повторной проверки. 
    
Очень важно вести диалог и обсуждать те пункты, которые у тебя могли не получиться, поэтому, ты можешь оставлять свои комментарии, и мы обязательно во всем разберемся! Самое главное - выделяй свои комментарии удобным для тебя способом (цвет, рамка), но так, чтобы они отчетлитво отличались от моих комментариев. Так мы ни в чем не запутаемся)

Свои комментарии я буду выделять так: <font color='green'>зеленый</font>, <font color='gold'>желтый</font> и <font color='red'>красный</font> цвет, например:

<br/>

<div class="alert alert-success">
<h2> Комментарий ревьюера V1<a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Этот комментарий я буду оставлять, если твое решение оптимально для данного шага.
</div>

<br/>

<div class="alert alert-warning">
    <h2> Комментарий ревьюера V1<a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> Это не означает, что ты совершил ошибку. Такими комментариями я буду отмечать места, где ты справился с задачей, но её можно решить ещё лучше. Также в таких комментариях я буду оставлять для тебя другие советы.
</div>


<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера V1<a class="tocSkip"></h2>

    
<b>На доработку🤔:</b>
Если решение, которое ты применил на определенном шаге, требует значительных изменений, то я буду оставлять такой комментарий. Тебе нужно будет подумать еще раз как правильным образом решить данный этап. Если ревью содержит такие комментарии, то я не смогу принять проект с первого раза.
</div>
    
P.S. В оглавлении комментария ты увидишь `V1` (`V1`, `V2`, `V3`) - это версия моего комментария, которая понадобится нам с тобой, чтобы не запутаться, если потребуется оставить тебе комментарии после первой проверки.

# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [22]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')
    
print(data.info())    # есть пропуски в столбцах 'days_employed' и 'total_income '


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB
None


In [23]:
#  Проверим столбцы 'children' и 'debt' на уникальные значения
print(data['children'].unique())
print(data['debt'].unique())
 
# Удалим строки в столбце 'children' с аномальными значениями и проверим ещё раз. 

data = data[(data['children'] != -1) & (data['children'] != 20)]
# print(data['children'].unique())



[ 1  0  3  2 -1  4 20  5]
[0 1]


In [24]:
# Сгруппируем датафрейм по количеству детей в семье. Узнаем количество заёмщиков в каждой группе.
# Узнаем количество заёмщиков, которые уже имели задолженность по возврату кредитов. 
# Значение столбца 'debt' = 0 (нет задолженностей), 'debt' = 1 (есть задолженности)
# children_grouped = data.groupby('children').agg({'debt': ['count', 'sum']}) 
# print(children_grouped)
# Рассчитаем конверсию для потенциальных должников среди заёмщиков в каждой группе.
# debt_children_percent = data.groupby('children')['debt'].sum() / data.groupby('children')['debt'].count()*100
# print(debt_children_percent)

# Изменила код

data_pivot_children = data.pivot_table(index=['children'], values=['debt'], aggfunc=['sum','count','mean'])

print(data_pivot_children)


           sum  count      mean
          debt   debt      debt
children                       
0         1063  14149  0.075129
1          444   4818  0.092154
2          194   2055  0.094404
3           27    330  0.081818
4            4     41  0.097561
5            0      9  0.000000


<div class="alert alert-warning">
    <h2> Комментарий ревьюера V1<a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> Вижу, что ты хорошо справляешься с методом `groupby()`, стоило объединить верхние две ячейки в одну сводную таблицу, чтобы легче было анализировать информацию по группам. Также ты можешь воспользоваться методом `pivot_table()`

<div class="alert alert-success">
<h2> Комментарий ревьюера V2<a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Отлично построила сводную таблицу! Согласись, что по ней гораздо удобнее анализировать группы клиентов.
</div>

**Новый Вывод:** 
- Для данного исследования конверсия в пределах 1% имеет значение, так как речь идет о возврате заемщиками банку значительных денежных сумм в срок.
- минимальную конверсию 7,5 % имеет группа заёмщиков без детей.
- конверсия в группе с 3-мя детьми составляет 8,2 %.
- группы заемщиков, имеющие 1- 2 ребенка показывают конверсию 9,2 и 9,4 % соответственно, что выше, чем в группе заёмщиков без    детей или в группе с 3-мя детьми.
- группы заемщиков имеющие 4 ребенка показывают наибольшую конверсию среди заемщиков - 9,8 %.
- в группе заёмщиков с 5 детьми не выявлены заемщики, которые не погашают кредит в срок, то есть конверсия 0 %. Но имеющиеся у нас данные    не позволяют сделать вывод о том, что заёмщики данной группы более добросовестные, чем остальные. Так как количество заещиков с 5 детьми достаточно мало, всего 9 человек, данной выборки недостаточно, чтобы сделать вывод по данной группе.

**Предыдущий вывод:**
- Группы заёмщиков с количеством детей от 0 до 4 имеют примерно одинаковый процент заёмщиков, которые не погашают кредит в срок     - 7-9 %. При этом процент недобросовестных заёмщиков, не имеющих детей в данной группе ниже, чем у остальных. 
- В группе заёмщиков с 5 детьми не выявлены заемщики, которые не погашают кредит в срок, то есть 0 %. Но имеющиеся у нас данные    не позволяют сделать вывод о том, что заёмщики данной группы более добросовестные, чем остальные. Так как количество заещиков    с 5 детьми достаточно мало (всего 9 человек), данной выборки недостаточно, чтобы сделать вывод по данной группе. 
- Возврат кредита в срок практически не зависит от количеством детей у заёмщика

<div class="alert alert-warning">
    <h2> Комментарий ревьюера V1<a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> Молодец, что отметила недостаточное количество наблюдений в группе клиентов с 5 детьми. Однако зависимость есть, клиенты без детей все таки реже допускают просрочки. Отличие на несколько % между группами в полне вероятно может быть значимым, поэтому делать вывод об отсутствии зависимости в данном случае неправильно.

<div class="alert alert-success">
<h2> Комментарий ревьюера V2<a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Абсолютно верно, так держать!
</div>

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [25]:
#  Проверим столбец 'family_status' на уникальные значения.

print(data['family_status'].unique())


['женат / замужем' 'гражданский брак' 'вдовец / вдова' 'в разводе'
 'Не женат / не замужем']


In [26]:
# Сгруппируем датафрейм по уникальным значениям в столбце 'family_status'.
# Узнаем количество заёмщиков в каждой группе.
# Узнаем количество заёмщиков, которые уже имели задолженность по возврату кредитов.
# Значение столбца 'debt' = 0 (нет задолженностей), 'debt' = 1 (есть задолженности).

#family_status_grouped = data.groupby('family_status').agg({'debt': ['count', 'sum']}) 
#print(family_status_grouped)

# Рассчитаем конверсию для потенциальных должников среди заёмщиков в каждой группе.
#debt_family_status_percent = data.groupby('family_status')['debt'].sum() / data.groupby('family_status')['debt'].count()*100
#print(debt_family_status_percent.sort_values())

# Изменила код

data_family_status_pivot = data.pivot_table(index=['family_status'], values=['debt'], aggfunc=['sum','count','mean'])

print(data_family_status_pivot) # Подскажи, нужно ли сортировать результат по возрастанию столбца mean debt
# и, если да, то как это сделать? 
# print(data_family_status_pivot.sort.value((by=' ????  '))

                       sum  count      mean
                      debt   debt      debt
family_status                              
Не женат / не замужем  273   2799  0.097535
в разводе               84   1189  0.070648
вдовец / вдова          63    952  0.066176
гражданский брак       385   4160  0.092548
женат / замужем        927  12302  0.075354


<div class="alert alert-warning">
    <h2> Комментарий ревьюера V1<a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> Также стоит объединить верхние ячейки в одну сводную таблицу.

<div class="alert alert-success">
<h2> Комментарий ревьюера V2<a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> И снова корректно построенная сводная таблица, видно, как ты успешно прокачиваешь свои навыки!
</div>

**Новый Вывод:** 
- анализ групп заемщиков по семейному положению показал? что наиболее надежными заемщиками является группа, имеющая семейный статус "вдовец / вдова". Конверсия составляет 6.7 %.
- далее по надежности с точки зрения возврата кредита в срок следуют группы " в разводе" и "женат / замужем" с конверсией 7.0 % и 7.5 % соответственно.
- наименее нажедными являются заёмщики в группах "гражданский брак" и "Не женат / не замужем"  с конверсией 9.3 % и 9.8 % соответственно.

**Предыдущий вывод:**
* анализ групп заемщиков по семейному положению показал, что в каждой группе имеется примерно одинаковый процент заёмщиков, которые не погашают кредит в срок  - 6-9 %.
* при этом более надежными заемщиками являются группы, имеющие семейный статус "вдовец / вдова"  и " в разводе".
  менее надежными заемщиками являются группы, имеющие семейный статус "гражданский брак"  и "Не женат / не замужем".
* Возврат кредита в срок практически не зависит от семейного положения заёмщика.

<div class="alert alert-warning">
    <h2> Комментарий ревьюера V1<a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> Опять же различия между группами есть, мы делаем выводы о разнице между ними, следовательно отличие на несколько процентов уже говорит о том, что некоторые группы клиентов реже допускают просрочки. Подсказка: если внимательно посмотреть на группы, то можно выделить 2 класса.

<div class="alert alert-success">
<h2> Комментарий ревьюера V2<a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Все верно, если посмотреть на группы, то можно выявить закономерность: клиенты, которые когда либо были в браке, реже допускают просрочки по кредитам.
</div>

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [27]:
# Заполним пропуски в столбце 'total_income' медианными значениями для каждого типа занятости income_type.

print('пропуски до', data['total_income'].isna().sum())

for person in data['income_type'].unique(): 
    median = data.loc[data['income_type'] == person, 'total_income'].median()
    #print(person, median)
    data.loc[(data['income_type'] == person) & (data['total_income'].isna()), 'total_income'] = median
print('пропуски после', data['income_type'].isna().sum())  


пропуски до 2162
пропуски после 0


In [28]:
# Используем функцию categorize_income(income) и столбец 'total_income_category' с категориями по уровню дохода:
#0–30000 — 'E';
#30001–50000 — 'D';
#50001–200000 — 'C';
#200001–1000000 — 'B';
#1000001 и выше — 'A'.
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass
data['total_income_category'] = data['total_income'].apply(categorize_income)
# print(data[['total_income','total_income_category']]. tail(10))
# print(data.tail(10))

In [29]:
# Сгруппируем датафрейм по уникальным значениям в столбце 'total_income_category'.
#Узнаем количество заёмщиков в каждой группе.
#Узнаем количество заёмщиков, которые уже имели задолженность 
# по возврату кредитов. Значение столбца 'debt' = 0 (нет задолженностей), 'debt' = 1 (есть задолженности)

#total_income_category_grouped = data.groupby('total_income_category').agg({'debt': ['count', 'sum']}) 
#print(total_income_category_grouped)
# Рассчитаем конверсию для потенциальных должников среди заёмщиков в каждой группе.
#debt_total_income_category_percent = data.groupby('total_income_category')['debt'].sum() / data.groupby('total_income_category')['debt'].count()*100
#print(debt_total_income_category_percent.sort_values())

# Изменила код

data_income_category_pivot = data.pivot_table(index=['total_income_category'], values=['debt'], aggfunc=['sum','count','mean'])

print(data_income_category_pivot)

                        sum  count      mean
                       debt   debt      debt
total_income_category                       
A                         2     25  0.080000
B                       354   5014  0.070602
C                      1353  15992  0.084605
D                        21    349  0.060172
E                         2     22  0.090909


<div class="alert alert-warning">
    <h2> Комментарий ревьюера V1<a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> Напоминаю про объединение в одну сводную таблицу.

<div class="alert alert-success">
<h2> Комментарий ревьюера V2<a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Все верно!
</div>

**Новый Вывод:** 

* количество клиентов в группах с уровнем дохода категории 'A'(1000001 и выше), 'D'(30001–50000) и 'E'(0–30000) достаточно мало, поэтому сложно сделать вывод о надежности заемщиков в этих группах на основании анализа: выборка может быть непоказательной.

* анализ групп с уровнем дохода категории 'B'(200001–1000000) и 'С'(50001–200000) показал, что наиболее надежными являются    заемщики имеющие уровень дохода категории 'B'(200001–1000000), конверсия в данной группе составляет 7.1 %. 
  Менее наждежные заемщики имеют уровень дохода категории 'С'(50001–200000), конверсия - 8.5 %.


**Предыдущий вывод:**
* анализ групп заемщиков по уровню дохода показал, что в каждой группе имеется примерно одинаковый процент заёмщиков, которые не погашают кредит в срок -  6-9 %.
* при этом более надежными заемщиками являются группы, имеющие уровень дохода категории 'D'(30001–50000) и 'B'(200001–1000000).
* Можно предположить, что менее надежными заемщиками являются группы, имеющие  уровень дохода категории 'E'(0–30000). Но количество заемщиков категории 'E'(0–30000), также как и заемщиков категории 'A'(1000001 и выше) достаточно мало, поэтому выборка может быть непоказательной.
* **Возврат кредита в срок практически не зависит от уровня дохода заёмщика.**

<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера V1<a class="tocSkip"></h2>

    
<b>На доработку🤔:</b>
Ты верно размышляешь, однак отмечу, что относительно друг друга группы отличаются по вероятности просрочки, даже если это отличие выражается в несколько процентов. Молодец, что учла недостаточное количество наблюдений в группах А и Е, как ты думаешь, можно ли анализировать группу D с группами В и С, достаточно ли в ней клиентов? Вывод о том, что зависимости нет, неправильный, посмотри на группы еще раз и постарайся сделать корректный вывод.
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера V2<a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Абсолютно верно! Из-за маленького количества наблюдений в выборках А, Е и D, мы не можем их анализировать. Если посмотреть на другие группы, то выявляется закономерность: чем выше доход, тем меньше вероятность просрочки.
</div>

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [30]:
# Используем функцию categorize_purpose и столбец 'purpose_category' со следующими категориями:
#'операции с автомобилем',
#'операции с недвижимостью',
#'проведение свадьбы',
#'получение образования'.
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'
data['purpose_category'] = data['purpose'].apply(categorize_purpose)    
# print(data[['purpose','purpose_category']]. head(10)) 

In [31]:
# Сгруппируем датафрейм по уникальным значениям в столбце 'purpose_category'.
#Узнаем количество заёмщиков в каждой группе.
#Узнаем количество заёмщиков, которые уже имели задолженность 
# по возврату кредитов. Значение столбца 'debt' = 0 (нет задолженностей), 'debt' = 1 (есть задолженности)
# purpose_category_grouped = data.groupby('purpose_category').agg({'debt': ['count', 'sum']}) 
#print(purpose_category_grouped)
# Рассчитаем конверсию для потенциальных должников среди заёмщиков в каждой группе.
#debt_purpose_category_percent = data.groupby('purpose_category')['debt'].sum() / data.groupby('purpose_category')['debt'].count()*100
#print(debt_purpose_category_percent.sort_values())

# Изменила код

data_purpose_category_pivot = data.pivot_table(index=['purpose_category'], values=['debt'], aggfunc=['sum','count','mean'])
print(data_purpose_category_pivot)

                          sum  count      mean
                         debt   debt      debt
purpose_category                              
операции с автомобилем    400   4288  0.093284
операции с недвижимостью  780  10780  0.072356
получение образования     369   3997  0.092319
проведение свадьбы        183   2337  0.078306


<div class="alert alert-warning">
    <h2> Комментарий ревьюера V1<a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> Здесь тоже было бы гораздо удобнее анализировать данные, если бы они были в одной сводной таблице.

<div class="alert alert-success">
<h2> Комментарий ревьюера V2<a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Молодец, что построила сводную таблицу.
</div>

**Новый Вывод:** 
* анализ групп заемщиков по цели кредита показал, что наиболее надежные заемщики - те, кто берет гредит на операции с недвижимостью и проведение свадьбы. Конверсия для данных групп составляет 7.2 % и 7.8 %. 
* просрочки допускают чаще заемщики, которые берут кредит на получение образования  или  операции с автомобилем. Конверсия для данных групп составляет 9.2 % и 9.3 %. 


**Предыдущий вывод:**
* анализ групп заемщиков по цели кредита показал, что в каждой группе имеется примерно одинаковый процент заёмщиков, которые не погашают кредит в срок - 7-9 %.
* при этом более надежные заемщики - те, кто берет гредит на операции с недвижимостью и  проведение свадьбы, менее надежные заемщики берут кредит на получение образования  или  операции с автомобилем. 
* Возврат кредита в срок практически не зависит от цели кредита.

<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера V1<a class="tocSkip"></h2>

    
<b>На доработку🤔:</b>
Аналогичный комментарий о том, что разница в несколько процентов может быть значимой, без дополнительных проверок нельзя её отвергнуть. Что касается кредитов на недвижимость и свадьбы, ты правильно отметила, что просрочки по ним допускаются реже.
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера V2<a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Теперь вывод полностью соответствует действительности. Полностью с ним согласен.
</div>

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* Пропуски выявлены одновременно в столбцах в "days_employed" и "total_income". Стаж работы и доход - это основные параметры, которые рассматриват банк при выдаче кредитов. Если бы заемщики не внесли эти данные намеренно, они бы не попали в базу банка и не получили кредит. Скорее всего данные не попали в базу по технической ошибке. 

<div class="alert alert-success">
<h2> Комментарий ревьюера V1<a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Все верно, эти причины вполне могут возникнуть в реальной жизни.
</div>

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* пропуски количественных значений будет правильно заполнить медианным значением, а не средним значением, так как среди количественных переменных в таблице могут быть "лидеры" - значения, значительно отличающиеся от остальных. Такие "лидеры" значительно смещают среднее значение и искажают результат. 

<div class="alert alert-success">
<h2> Комментарий ревьюера V1<a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Абсолютно верный ответ!
</div>

### Шаг 4: общий вывод.

**Новый Вывод:** 

* Для данного исследования конверсия в пределах 1% имеет значение, так как речь идет о возврате заемщиками банку значительных денежных сумм в срок.

* Наименьшую конверсию показывает такие категории заемщиков как **"вдовец / вдова",  "в разводе"**  и  **"женат / замужем",   неимеющие детей**, с доходом категории **'B'(200001–1000000)** на такие цели как **'операции с недвижимостью'**  или  **'проведение свадьбы'**.

* Наибольшую конверсию показывает такие категории заемщиков как **'гражданский брак'**  и  **'Не женат / не замужем',  имеющие детей**, с доходом ктегории **'С'(50001–200000)** на такие цели как **'получение образования'**  или  **'операции с автомобилем'**.


**Предыдущий вывод:**
* Анализ таблицы показал, что такие показатели, 
* Наименьшую конверсию показывает категория заемщиков  "вдовец / вдова" и "в разводе", неимеющие детей, с доходом 30001–50000('D') или 200001–1000000('B') на такие цели как 'операции с недвижимостью' или 'проведение свадьбы'
* Наибольшую конверсию показывает категория заемщиков  'гражданский брак' и 'Не женат / не замужем', имеющие детей, с доходом 0–30000 ('E') или 50001–200000 ('C') на такие цели как 'получение образования' или 'операции с автомобилем'

<div class="alert alert-warning">
    <h2> Комментарий ревьюера V1<a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> Стоит несколько доработать итоговый вывод с учетом моих комментариев выше. В целом ты хорошо описала результаты исследования, молодец, что выделила промежуточные выводы в отдельные выводы.

<div class="alert alert-success">
<h2> Комментарий ревьюера V2<a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Потрясующий вывод, все четко и структурировано. Такой вывод можно отправлять в реальный кредитный отдел банка!
</div>

<div style="border:solid Chocolate 2px; padding: 40px">

**Общий вывод по проекту**

Спасибо за твой проект! Я уверен, что ты приложила все усилия, чтобы сделать этот проект. Это видно по твоей работе, а сформулированные выводы отражают, насколько успешно ты осваиваешь материал. 

**Отмечу отдельные положительные моменты проекта🙂**:
 
* Ты полностью следовала всем этапам выполнения задания и справилась почти со всем. Хорошая работа!
* Отдельно выделю некоторые выводы, которые ты смогла сформулировать абсолютно точно!
* У тебя опрятный и читаемый код. Возможно это звучит банально, однако при работе с чужим кодом это крайне важно)


**На что стоит обратить внимание🤓**:
    
* Важно доработать итоговый вывод.
* Проработать все промежуточные выводы с учетом моих комментариев.
* Остальные рекомендации ты найдешь в моих комментариях!


**Желаю удачи и жду твой проект на повторное ревью!**😉

<div style="border:solid Chocolate 2px; padding: 40px">

Ты отлично справилась с данной работой! Видно, как ты успешно осваиваешь материал, применяешь полученные знания на практике. Мне было очень приятно проверять твою работу, надеюсь, мои комментарии и советы были для тебя полезными.

**Желаю удачи в последующем обучении и исследованиях!**😉